staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -446,28 +451,106 @@ func (p *patcher) applyAdmission(ctx context.Context, patchedObject runtime.Obje
+		newObjVersioned, err := convert(newObj)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert new object to proper version: %v", err)
+		}
+		oldObjVersioned, err := convert(oldObj)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert old object to proper version: %v", err)
+		}
+
+		managed, err := apply.DecodeObjectManagedFields(newObj)
+		if err != nil {
+			return nil, err
+		}
+
+		newObjTyped, err := typeConverter.ObjectToTyped(newObjVersioned)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create typed new object: %v", err)
+		}
+		oldObjTyped, err := typeConverter.ObjectToTyped(oldObjVersioned)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create typed old object: %v", err)
+		}
+
+		managed, err = updater.Update(oldObjTyped, newObjTyped, managed, manager)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update ManagedFields: %v", err)
+		}
+
+		if err := apply.EncodeObjectManagedFields(newObj, managed); err != nil {
+			return nil, err
+		}
+
+		return newObj, nil
+	}
+}
+
 // patchResource divides PatchResource for easier unit testing
 func (p *patcher) patchResource(ctx context.Context, scope RequestScope) (runtime.Object, bool, error) {
+	var (
+		converter apply.TypeConverter
+		err       error
+	)
+
+	if scope.OpenAPIModels != nil {
+		converter, err = apply.NewTypeConverter(scope.OpenAPIModels)
+		if err != nil {
+			return nil, false, err
+		}
+	}
+
+	updater := merge.Updater{
+		Converter: apply.NewVersionConverter(converter, p.unsafeConvertor),
+	}
+
+	p.applyManagedFields = noOpTransform
 	p.namespace = request.NamespaceValue(ctx)
 	switch p.patchType {
 	case types.JSONPatchType, types.MergePatchType:
 		p.mechanism = &jsonPatcher{patcher: p}
+		if scope.OpenAPIModels != nil {
+			p.applyManagedFields = makeManagedFieldsUpdater(converter, newConvertor(p.unsafeConvertor, p.kind.GroupVersion()), updater, "jsonpatch")
+		}
 	case types.StrategicMergePatchType:
 		schemaReferenceObj, err := p.unsafeConvertor.ConvertToVersion(p.restPatcher.New(), p.kind.GroupVersion())
 		if err != nil {
 			return nil, false, err
 		}
 		p.mechanism = &smpPatcher{patcher: p, schemaReferenceObj: schemaReferenceObj}
-	// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
+		if scope.OpenAPIModels != nil {
+			p.applyManagedFields = makeManagedFieldsUpdater(converter, newConvertor(p.unsafeConvertor, p.kind.GroupVersion()), updater, "smpatch")
+		}
+		// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
 	case types.ApplyPatchType:
-		p.mechanism = &applyPatcher{patcher: p, models: scope.OpenAPIModels}
+		if scope.OpenAPIModels == nil {
+			klog.Warningf("No OpenAPI for: %v", scope.Kind)
+			return nil, false, nil
+		}
+		p.mechanism = &applyPatcher{patcher: p, converter: converter, updater: updater}
```

Comment:
It looks like the `if scope.OpenAPIModels == nil` condition is repeated many times. Can we change this code just a little bit so that it doesn't happen? Also if you add a condition on the serversideapply flag, it's going to be even worse.